### PR TITLE
S2U-36 5.2.1.1 Announcements - Ability to assign announcements to a role

### PR DESCRIPTION
**Pull request by JuanDavid102**

### CODE REVIEW

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 1]
```diff
@@ -1679,7 +1758,18 @@ protected String buildReviseAnnouncementContext(VelocityPortlet portlet, Context
...
+		try {
+			Site site = SiteService.getSite(siteId);
+			Role[] siteRoles = site.getRoles().toArray(new Role[site.getRoles().size()]);
+			List<String> siteRolesIds = new ArrayList<String>();
+			for (Role role: siteRoles){
+				siteRolesIds.add(role.getId());
+			}
+			context.put("siteRolesIds", siteRolesIds);
+		} catch (IdUnusedException ex) {
+			log.error(ex.getMessage());
```
Comment (jesusmmp) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:1771]: More description of the error?

File: library/src/morpheus-master/bootstrap-sass-3.3.7/assets/stylesheets/bootstrap/_forms.scss [hunk 2]
```diff
@@ -246,7 +246,8 @@ input[type="search"] {
 }
 
 .radio + .radio,
-.checkbox + .checkbox {
+.checkbox + .checkbox,
+.reduction-top {
```
Comment (jesusmmp) [inline library/src/morpheus-master/bootstrap-sass-3.3.7/assets/stylesheets/bootstrap/_forms.scss:250]: I'll try avoid margin-top negatives...
Could we reduce the margin-bottom from the top element?

Comment (JuanDavid102) [reply library/src/morpheus-master/bootstrap-sass-3.3.7/assets/stylesheets/bootstrap/_forms.scss:250]: I don't know, i will try now

Comment (JuanDavid102) [reply library/src/morpheus-master/bootstrap-sass-3.3.7/assets/stylesheets/bootstrap/_forms.scss:250]: Yes, it works. In some minutes, i will send the new "version" with that changes

File: announcement/announcement-api/api/src/bundle/announcement_eu.properties [hunk 3]
```diff
@@ -329,3 +329,18 @@ perm-annc.delete.any=Ohar guztiak ezabatu
 perm-annc.delete.own=Norberaren oharrak ezabatu
 perm-annc.read.drafts=Oharren zirriborroak irakurri
 perm-annc.all.groups=Atal guztien oharrak eskuragarri izan
+
+### S2U-36 Announcements - Ability to assign announcements to a role
```
Comment (mpellicer) [inline announcement/announcement-api/api/src/bundle/announcement_eu.properties:333]: @PabloSanRoman could you help with the translations of the feature? Thanks!

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/entityprovider/AnnouncementEntityProviderImpl.java [hunk 4]
```diff
@@ -277,9 +278,32 @@ private List<?> getAnnouncements(String siteId, Map<String,Object> params, boole
...
+				showAnnouncement = true;
+			} else {
+				String ownerId = a.getAnnouncementHeader().getFrom().getId();
+				String userId = sessionManager.getCurrentSessionUserId();
+				if (ownerId.equals(userId)) {
+					showAnnouncement = true;
+				} else {
+					Member actualMember = site.getMember(userId);
+					String memberRole = actualMember.getRole().getId();
+					for (String selectedRole : selectedRoles.split(", ")) {
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/entityprovider/AnnouncementEntityProviderImpl.java:295]: Good job here!

I kinda don't like this split with the space, let's improve it.

Comment (mpellicer) [reply announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/entityprovider/AnnouncementEntityProviderImpl.java:295]: if (securityService.isSuperUser() || selectedRoles == null) {
showAnnouncement = true;
} else {
}

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/entityprovider/AnnouncementEntityProviderImpl.java [hunk 5]
```diff
@@ -277,9 +278,32 @@ private List<?> getAnnouncements(String siteId, Map<String,Object> params, boole
 	
 		for (Message m : announcements) {
 			AnnouncementMessage a = (AnnouncementMessage)m;
+			boolean showAnnouncement = false;
+			String selectedRoles = ((String) a.getProperties().get("selectedRoles"));
+			if (securityService.isSuperUser()) {
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/entityprovider/AnnouncementEntityProviderImpl.java:283]: if (securityService.isSuperUser() || selectedRoles == null) {
showAnnouncement = true;
} else {
}

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 6]
```diff
@@ -466,6 +471,10 @@ static String getAnnouncementRange(AnnouncementMessage a)
 		{
 			return rb.getString("gen.public");
 		}
+		else if (a.getProperties().getProperty(SELECTED_ROLES_PROPERTY) != null)
+		{
+			return a.getProperties().getProperty(SELECTED_ROLES_PROPERTY).replaceAll("[\\[\\]]", "");
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:476]: Could you explain this replaceAll to me?

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 7]
```diff
@@ -694,11 +703,42 @@ else if (view.equals(VIEW_MODE_PUBLIC))
...
 						else
 						{
 							messages = getMessages(channel, null, true, state, portlet);
 						}
+						List<AnnouncementWrapper> finalMessages = new ArrayList<>();
+						for (AnnouncementWrapper message : messages) {
+							site = SiteService.getSite(EntityManager.newReference(message.getReference()).getContext());
+							String selectedRoles = (String) message.getProperties().get(SELECTED_ROLES_PROPERTY);
+							boolean addAnnounMessage = false;
+							if (selectedRoles == null) {
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:720]: Shall we check the superUser here ?

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 8]
```diff
@@ -694,11 +703,42 @@ else if (view.equals(VIEW_MODE_PUBLIC))
...
+								addAnnounMessage = true;
+							} else {
+								String[] selectedRolesArray = selectedRoles.replaceAll("[\\[\\]]", "").replaceAll(", ", ",").split(",");
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:723]: This is still weird, replaceAll, replace ', ', and an split.

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 9]
```diff
@@ -694,11 +703,42 @@ else if (view.equals(VIEW_MODE_PUBLIC))
...
+								Member actualMember = site.getMember(userDirectoryService.getCurrentUser().getId());
+								String memberRole = actualMember.getRole().getId();
+								for (String selectedRole : selectedRolesArray) {
+									if (selectedRole.equals(memberRole)) {
+										addAnnounMessage = true;
+									}
+								}
+								if (actualMember.getUserEid().equals(message.getAnnouncementHeader().getFrom().getDisplayId())) {
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:731]: It should be getId ?

the variable should be currentMember

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java (hunk 7 above)
Comment (JuanDavid102) [reply announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:720]: Yes

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/entityprovider/AnnouncementEntityProviderImpl.java (hunk 5 above)
Comment (mpellicer) [reply announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/entityprovider/AnnouncementEntityProviderImpl.java:283]: Let's be consistent with the action,

String ownerId = a.getAnnouncementHeader().getFrom().getId();
String userId = sessionManager.getCurrentSessionUserId();
boolean isOwner = ownerId.equals(userId);

if (isOwner || securityService.isSuperUser() || selectedRoles == null) {
} else {
}

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java (hunk 7 above)
Comment (mpellicer) [reply announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:720]: Let's be consistent with the entity provider.

String ownerId = a.getAnnouncementHeader().getFrom().getId();
String userId = sessionManager.getCurrentSessionUserId();
boolean isOwner = ownerId.equals(userId);

if (isOwner || securityService.isSuperUser() || selectedRoles == null) {
} else {
}

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 10]
```diff
@@ -1579,6 +1653,11 @@ protected String buildPreviewContext(VelocityPortlet portlet, Context context, R
 		}
 		}
 		
+		String annToRoles = state.getTempAnnounceToRoles();
+		String allRolesString = "";
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:1657]: Is this unused?

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 11]
```diff
@@ -1716,6 +1806,10 @@ else if (channel.getGroupsAllowAddMessage().size() > 0)
 							// to group otherwise
 							context.put("announceTo", "groups");
 						}
+					} else {
+						state.setTempAnnounceToRoles(
+							state.getEdit().getProperties().getProperty(SELECTED_ROLES_PROPERTY).replaceAll("[\\[\\]]", "").replaceAll(", ", ",")
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:1811]: Same concern with replaces and replaceAll.

File: announcement/announcement-tool/tool/src/webapp/vm/announcement/chef_announcements-revise.vm [hunk 12]
```diff
@@ -69,6 +69,20 @@ function resizeFrame(){
             includeSelectAllOption: true,
             enableCaseInsensitiveFiltering: true
         });
+
+		$('#selectedRoles').multiselect(
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/webapp/vm/announcement/chef_announcements-revise.vm:73]: Spacing, minor.

File: library/src/morpheus-master/bootstrap-sass-3.3.7/assets/stylesheets/bootstrap/_forms.scss (hunk 2 above)
Comment (mpellicer) [reply library/src/morpheus-master/bootstrap-sass-3.3.7/assets/stylesheets/bootstrap/_forms.scss:250]: Same, try to do not modify the styles here!

File: portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/SkinnableCharonPortal.java [hunk 13]
```diff
@@ -1910,6 +1910,7 @@ public void includeLogin(PortalRenderContext rcontext, HttpServletRequest req,
 					message = rloader.getFormattedMessage("sit_return", impersonatorDisplayId);
 					rcontext.put("impersonatorDisplayId", impersonatorDisplayId);
 				}
+				rcontext.put("userDisplayName", Validator.escapeHtml(UserDirectoryService.getCurrentUser().getDisplayName()));
```
Comment (mpellicer) [inline portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/SkinnableCharonPortal.java:1913]: Do we still need this change?

File: portal/portal-render-engine-impl/impl/src/webapp/vm/morpheus/includeStandardHead.vm [hunk 14]
```diff
@@ -36,6 +36,7 @@
                     "userType": "$!{userType}",
                     "siteRole": "$!{userSiteRole}",
                     "impersonatorDisplayId": "$!{impersonatorDisplayId}",
+                    "userDisplayName": "$!{userDisplayName}",
```
Comment (mpellicer) [inline portal/portal-render-engine-impl/impl/src/webapp/vm/morpheus/includeStandardHead.vm:39]: Do we still need this change?

**Review by mpellicer — CHANGES_REQUESTED** [review]
Much better @JuanDavid102, we're closer!

File: portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/SkinnableCharonPortal.java (hunk 13 above)
Comment (JuanDavid102) [reply portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/SkinnableCharonPortal.java:1913]: No, the next version doesn't have it

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java (hunk 9 above)
Comment (JuanDavid102) [reply announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:731]: Ok

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 15]
```diff
@@ -694,11 +710,38 @@ else if (view.equals(VIEW_MODE_PUBLIC))
...
 						else
 						{
 							messages = getMessages(channel, null, true, state, portlet);
 						}
+						List<AnnouncementWrapper> finalMessages = new ArrayList<>();
+						for (AnnouncementWrapper message : messages) {
+							site = SiteService.getSite(EntityManager.newReference(message.getReference()).getContext());
+							Collection selectedRoles = message.getProperties().getPropertyList(SELECTED_ROLES_PROPERTY);
+							boolean showAnnounMessage = false;
+							if (selectedRoles == null || m_securityService.isSuperUser()) {
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:727]: Same as above, do it in one if.

boolean isOwner = (userDirectoryService.getCurrentUser().getId().equals(message.getAnnouncementHeader().getFrom().getId());

if (isOwner ||isAdmin || selectedRoles == null

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java [hunk 16]
```diff
@@ -694,11 +710,38 @@ else if (view.equals(VIEW_MODE_PUBLIC))
...
+								showAnnounMessage = true;
+							} else if (userDirectoryService.getCurrentUser().getId().equals(message.getAnnouncementHeader().getFrom().getId())) {
+								showAnnounMessage = true;
+							} else {
+								Member currentMember = site.getMember(userDirectoryService.getCurrentUser().getId());
+								String memberRole = currentMember.getRole().getId();
+								for (Iterator i = selectedRoles.iterator(); i.hasNext();) {
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/tool/AnnouncementAction.java:734]: Could we use for (String selectedRole : selectedRoles) ?

File: announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/entityprovider/AnnouncementEntityProviderImpl.java [hunk 17]
```diff
@@ -277,9 +280,26 @@ private List<?> getAnnouncements(String siteId, Map<String,Object> params, boole
...
+			boolean showAnnouncement = false;
+			Collection selectedRoles = a.getProperties().getPropertyList("selectedRoles");
+			String ownerId = a.getAnnouncementHeader().getFrom().getId();
+			boolean isOwner = ownerId.equals(currentUserId);
+			if (isOwner || securityService.isSuperUser() || selectedRoles == null) {
+				showAnnouncement = true;
+			} else {
+				Member currentMember = site.getMember(currentUserId);
+				String memberRole = currentMember.getRole().getId();
+				for (Iterator i = selectedRoles.iterator(); i.hasNext();) {
```
Comment (mpellicer) [inline announcement/announcement-tool/tool/src/java/org/sakaiproject/announcement/entityprovider/AnnouncementEntityProviderImpl.java:292]: Could we use for (String selectedRole : selectedRoles) ?

**Review by mpellicer — APPROVED** [review]

File: announcement/announcement-tool/tool/src/webapp/vm/announcement/chef_announcements.vm [hunk 18]
```diff
@@ -345,6 +346,8 @@
 					$tlang.getString("gen.therearepublic")
 				#elseif ($!view == "view.bygroup")
 					$tlang.getString("gen.therearegroup")
+				#elseif ($!view == "view.byrole")
+					$tlang.getString("gen.therearegroup")
```
Comment (susowrc) [inline announcement/announcement-tool/tool/src/webapp/vm/announcement/chef_announcements.vm:350]: @mpellicer Shouldn't this line be $tlang.getString("gen.therearerole") ?

Comment (mpellicer) [reply announcement/announcement-tool/tool/src/webapp/vm/announcement/chef_announcements.vm:350]: True! Thanks for the review!

I don't see where is happening, probably not visible, but you're right!

image

image


M

Comment (mpellicer) [reply announcement/announcement-tool/tool/src/webapp/vm/announcement/chef_announcements.vm:350]: @susowrc thanks again and good catch, this commit fixes the issue!

https://github.com/sakaiproject/sakai/commit/26b24bd010b053c0b5d185db5388d18c4ad5f26d

Comment (susowrc) [reply announcement/announcement-tool/tool/src/webapp/vm/announcement/chef_announcements.vm:350]: you’re welcome!

**Pull request merged**